### PR TITLE
feat(alpenglow): introduce alpenclock primitives to bank

### PR DESCRIPTION
**Pull request by ksn6**
#### Problem
This PR upstreams the primitives needed within `Bank` to implement the Alpenglow clock (i.e., the Alpenclock).

See https://github.com/solana-foundation/solana-improvement-documents/pull/363 for details on how the Alpenclock works.

#### Summary of Changes
In this PR, we introduce:

- An off-curve account used to track the nanosecond-resolution Alpenclock
- A getter, which obtains the value of the Alpenclock from this account
- A setter, which (1) updates the off-curve account as well as (2) the clock sysvar

We feature-gate the setter to avoid accidental invocation prior to launching Alpenglow.

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -2896,6 +2904,59 @@ impl Bank {
         })
     }
 
+    /// Update the clock sysvar from a block footer's nanosecond timestamp.
+    /// Also stores the nanosecond value for later retrieval via `get_nanosecond_clock`.
+    pub fn update_clock_from_footer(&self, time_nanos: i64) {
+        if !self.feature_set.is_active(&feature_set::alpenglow::id()) {
+            return;
+        }
```
Comment (ksn6) [inline runtime/src/bank.rs:2927]: Added these lines in to prevent accidental invocation.

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -2896,6 +2904,59 @@ impl Bank {
...
+
+        let time_secs = time_nanos / 1_000_000_000;
+
+        let parent_epoch = self.parent().map(|p| p.epoch());
+        let epoch_start_timestamp =
```
Comment (AshwinSekar) [inline runtime/src/bank.rs]: comparing to alpenglow https://github.com/anza-xyz/alpenglow/blob/c3ed24ca974497167b739f73866f55ce8131aa4b/runtime/src/bank.rs#L2063

confirming that we don't need special handling for genesis? it shouldn't have a block footer, just want to make sure it was intentionally dropped during the upstream

Comment (ksn6) [reply runtime/src/bank.rs]: yep, tried rewriting this logic to make it cleaner; slot 0 can never be alpenglow / genesis shouldn't have a block footer

that being said - I think I've changed my mind; I'll leave this logic in for now, since we'll want to get local cluster tests in at some point

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -2901,6 +2910,61 @@ impl Bank {
...
+
+        self.update_sysvar_account(&sysvar::clock::id(), |account| {
+            create_account(
+                &clock,
+                self.inherit_specially_retained_account_fields(account),
+            )
+        });
+
+        // Update Alpenglow clock
+        let alpenclock_size = wincode::serialized_size(&unix_timestamp_nanos).unwrap();
```
Comment (ksn6) [inline runtime/src/bank.rs]: In the previous change, I had this as just inspecting the number of bytes in `unix_timestamp_nanos`, but that's brittle; not ideal that this requires adding in a dependency, though `wincode` likely gets introduced at some point anyways.

File: runtime/src/bank.rs [hunk 4]
```diff
@@ -2901,6 +2910,61 @@ impl Bank {
...
+    /// Also stores the nanosecond value for later retrieval via `get_nanosecond_clock`.
+    pub fn update_clock_from_footer(&self, unix_timestamp_nanos: i64) {
+        if !self.feature_set.is_active(&feature_set::alpenglow::id()) {
+            return;
+        }
+
+        // On epoch boundaries, update epoch_start_timestamp
+        let unix_timestamp_s = unix_timestamp_nanos / 1_000_000_000;
+        let epoch_start_timestamp = match (self.slot, self.parent()) {
+            (0, _) => self.unix_timestamp_from_genesis(),
```
Comment (ksn6) [inline runtime/src/bank.rs:2939]: This line is basically just for tests.

Comment (AshwinSekar) [reply runtime/src/bank.rs:2939]: is this function actually called for tests for the genesis block? 
If we start a cluster with alpenglow enabled at genesis how is the initial value of the clock set? 

Just want to make sure I completely understand the corner cases before we upstream

File: runtime/src/bank.rs (hunk 3 above)
Comment (AshwinSekar) [reply runtime/src/bank.rs]: I believe `AccountSharedData::new_data` below is using bincode

Should we just use wincode instead?

File: runtime/src/bank.rs (hunk 4 above)
Comment (ksn6) [reply runtime/src/bank.rs:2939]: > is this function actually called for tests for the genesis block?

It's not; the genesis block's bank is created via `new_from_genesis`, which calls `update_clock` unconditionally. In `update_clock`, we have a check for whether `slot == 0`, and if that's the case, the clock is set to `self.unix_timestamp_from_genesis()`.

You're correct in that we can probably get rid of this first match statement, though I'd rather keep this here for completeness.

I'll add in a comment for clarity, though.

> Just want to make sure I completely understand the corner cases before we upstream

Writing this here for completeness - the rest of the banks are initialized via `_new_from_parent`, where there's this logic:

```rust
if new.get_alpenglow_genesis_certificate().is_none() {
    new.update_clock(Some(parent.epoch()));
}
```

During migration, this is the line that runs; concretely, on the zero'th slot (i.e., the zero'th Alpenglow slot), the genesis certificate should be set to `None`, meaning that we update the clock via the classic stake-weighted approach.

Later in the block, we then observe the footer, which updates the clock for the second time in the zero'th Alpenglow slot. So the clock is actually updated twice in the zero'th Alpenglow slot, when migrating from TowerBFT.

For Alpenglow slots 1+, the `new.update_clock` logic doesn't fire, because the Alpenglow genesis certificate is present; it's purely footer-based after that.

File: runtime/src/bank.rs [hunk 5]
```diff
@@ -2911,6 +2919,64 @@ impl Bank {
...
+                &clock,
+                self.inherit_specially_retained_account_fields(account),
+            )
+        });
+
+        // Update Alpenglow clock
+        let data = wincode::serialize(&unix_timestamp_nanos).unwrap();
+        let lamports = Rent::default().minimum_balance(data.len());
+        let mut alpenclock_acct = AccountSharedData::new(lamports, data.len(), &system_program::ID);
+        alpenclock_acct.set_data_from_slice(&data);
```
Comment (AshwinSekar) [inline runtime/src/bank.rs:2966]: no action for this PR, but since this `wincode::serialize -> AccountSharedData::set_data_from_slice ` has popped up a couple times maybe we make a new constructor for it

**Review by AshwinSekar — APPROVED** [review]

File: runtime/src/bank.rs (hunk 5 above)
Comment (ksn6) [reply runtime/src/bank.rs:2966]: sg - third time's the charm
